ui/pages/confirmations/utils/token.ts

```diff
@@ -1,32 +1,89 @@
 import { memoize } from 'lodash';
 import { Hex } from '@metamask/utils';
+import { AssetsContractController } from '@metamask/assets-controllers';
 import { getTokenStandardAndDetails } from '../../../store/actions';
 
+export type TokenDetailsERC20 = Awaited<
+  ReturnType<
+    ReturnType<AssetsContractController['getERC20Standard']>['getDetails']
+  >
+> & { decimalsNumber: number };
+
+export type TokenDetailsERC721 = Awaited<
+  ReturnType<
+    ReturnType<AssetsContractController['getERC721Standard']>['getDetails']
+  >
+>;
+
+export type TokenDetailsERC1155 = Awaited<
+  ReturnType<
+    ReturnType<AssetsContractController['getERC1155Standard']>['getDetails']
+  >
+>;
+
+export type TokenDetails =
+  | TokenDetailsERC20
+  | TokenDetailsERC721
+  | TokenDetailsERC1155;
+
 export const ERC20_DEFAULT_DECIMALS = 18;
 
-/**
- * Fetches the decimals for the given token address.
- *
- * @param {Hex | string} address - The ethereum token contract address. It is expected to be in hex format.
- * We currently accept strings since we have a patch that accepts a custom string
- * {@see .yarn/patches/@metamask-eth-json-rpc-middleware-npm-14.0.1-b6c2ccbe8c.patch}
- */
-export const fetchErc20Decimals = memoize(
-  async (address: Hex | string): Promise<number> => {
+export const parseTokenDetailDecimals = (
+  decStr?: string,
+): number | undefined => {
+  if (!decStr) {
+    return undefined;
+  }
+
+  for (const radix of [10, 16]) {
+    const parsedDec = parseInt(decStr, radix);
+    if (isFinite(parsedDec)) {
+      return parsedDec;
+    }
+  }
+  return undefined;
+};
+
+export const memoizedGetTokenStandardAndDetails = memoize(
```

Comment:
How long do we want to memoisation persists? 
If `getTokenStandardAndDetails` is eventually going for an asset API, if there is a network error occurring, in that case don't we want to ask again rather than returning an empty from memory?

Reply:
We submit this request to background and this I expect it to resolve with some value if that is possible.

Reply:
Sorry for missing the obvious but I am not sure if I understood why we memoise the background call at this point.

Reply:
Assume that we had an error on the call of `getTokenStandardAndDetails`, which could resolve on subsequent calls. Why we are persisting at that point?

Reply:
Well memoising is not added in this PR , it only refactored here. I think it so as to avoid calling controller method un-nessarily.